### PR TITLE
Munki3dev

### PR DESCRIPTION
**Pull request by mindovermiles262**
See https://github.com/munki/munki/pull/935

Adds colorized outputs to `makecatalogs` terminal output

### CODE REVIEW

File: code/client/munkilib/admin/makecatalogslib.py [hunk 1]
```diff
@@ -50,7 +49,9 @@ def hash_icons(repo, output_fn=None):
         icon_list.remove('_icon_hashes.plist')
     for icon_ref in icon_list:
         if output_fn:
-            output_fn("Hashing %s..." % (icon_ref))
+            output_fn(
+                colorize.OKGREEN + "Hashing " + colorize.ENDC +
+                "%s..." % (icon_ref))
```
Comment (gregneagle) [inline code/client/munkilib/admin/makecatalogslib.py]: If output_fn isn't printing to stdout -- like if it's appending to a log, or sending info to Managed Software Center, what would be the effect of adding these colorization codes?

**Comment by gregneagle** [comment]
See also https://github.com/munki/munki/wiki/Contributing-to-Munki#contributor-license-agreement

**Comment by mindovermiles262** [comment]
Currently the color codes would be included as text to the output/log. 

Would only including colorized output if output_fn=None be the way to solve this?

**Comment by gregneagle** [comment]
Maybe, but that adds a lot of if/then complexity to simple output lines.

**Comment by gregneagle** [comment]
Maybe you should consider _not_ being so aggressive with your colors and simply wrap the print fn with one that looks for lines starting with "WARNING" and "ERROR" (or whatever) and colorize those.

I get what you are trying to do here, but not yet convinced this won't trigger other issues.

**Comment by mindovermiles262** [comment]
Alternatively I can make colorize a function that checks if output_fn is STDOUT and adds the color codes if such

**Comment by gregneagle** [comment]
What if STDOUT is a pipe to another command and is not going to the terminal?
I'm pretty ignorant about all this.

**Comment by mindovermiles262** [comment]
STDOUT piped will include the color codes regardless of how it’s implemented. 

If you want to avoid this I think the best way would be to add a `—no-color` flag and change the colorize class to a function that checks for this flag

**Comment by gregneagle** [comment]
I think a concern I have is "What if people have scripts that are parsing the output of makecatalogs?" (to post things to Slack or generate emails, or other things) MWA2 parses the output. I worry that the additional of colorization codes could break these. So -- maybe my concerns are alleviated if colorized output is opt-in: a `--colorized-output` flag, perhaps, or an environment variable, or even a preference read from somewhere.

**Comment by gregneagle** [comment]
And I see we've hit upon the same basic idea, though I'd want it to behave the same as now by default.

**Comment by mindovermiles262** [comment]
Agree and understand. 

Let me take a deeper look to find the best place to add an opt-in variable and I’ll update this PR

**Comment by gregneagle** [comment]
I think some UNIX-y tools attempt to detect if STDOUT is connected to a tty-like object and only colorize the output if so, but I have zero idea how to do that.

**Comment by gregneagle** [comment]
This looks promising:
https://stackoverflow.com/questions/1077113/how-do-i-detect-whether-sys-stdout-is-attached-to-terminal-or-not

which suggests
```python
if sys.stdout.isatty():
    # You're running in a real terminal
else:
    # You're being piped or redirected
```

**Comment by gregneagle** [comment]
In any case, if the colorization is not-opt-in, we'll need to carefully communicate the change so people can test to make sure this doesn't break any tooling they've wrapped around `makecatalogs` output. I mostly worry about `if output.startswith("XXX"):`-style parsing.

**Comment by mindovermiles262** [comment]
I think setting an option/setting is the best way to go about this. That way nothing breaks but the feature is still available for those that want it

**Pull request closed without merging**